File: packages/nextalign/src/translate/translateGenes.cpp
#include "translateGenes.h"

#include <common/copy.h>
#include <common/debug_trace.h>
#include <common/safe_vector.h>
#include <fmt/format.h>
#include <nextalign/nextalign.h>

#include <map>
#include <string>
#include <string_view>

#include "../alphabet/nucleotides.h"
#include "../strip/stripInsertions.h"
#include "../utils/at.h"
#include "../utils/mapFind.h"
#include "./extractGene.h"
#include "./mapCoordinates.h"
#include "./translate.h"
#include "align/alignPairwise.h"
#include "align/alignmentParams.h"
#include "decode.h"
#include "detectFrameShifts.h"
#include "removeGaps.h"


void maskNucFrameShiftsInPlace(NucleotideSequence& seq,
  const safe_vector<InternalFrameShiftResultWithMask>& frameShifts) {
  for (const auto& frameShift : frameShifts) {
    auto current = frameShift.frameShift.nucRel.begin;
    const auto end = frameShift.frameShift.nucRel.end;
    while (current < end) {
      if (at(seq, current) != Nucleotide::GAP) {
        at(seq, current) = Nucleotide::N;
      }
      ++current;
    }
  }
}


template<typename Letter>
void fillRangeInplace(Sequence<Letter>& seq, const Range& range, Letter letter) {
  auto current = range.begin;
  const auto end = range.end;
  while (current < end) {
    at(seq, current) = letter;
    ++current;
  }
}

/**
 * Mask gaps in frame shifted regions of the peptide.
 * This region is likely misaligned, so these gaps added during peptide alignment don't make sense.
 */
void maskPeptideFrameShiftsInPlace(AminoacidSequence& seq,
  const safe_vector<InternalFrameShiftResultWithMask>& frameShifts) {
  for (const auto& frameShift : frameShifts) {
    const auto& gapsLeading = frameShift.frameShift.gapsLeading.codon;
    const auto& frameShiftBody = frameShift.frameShift.codon;
    const auto& gapsTrailing = frameShift.frameShift.gapsTrailing.codon;

    fillRangeInplace(seq, gapsLeading, Aminoacid::GAP);
    fillRangeInplace(seq, frameShiftBody, Aminoacid::X);
    fillRangeInplace(seq, gapsTrailing, Aminoacid::GAP);
  }
}

/** Converts frame shift internal representation to external representation  */
safe_vector<FrameShiftResult> toExternal(const safe_vector<InternalFrameShiftResultWithMask>& frameShiftResults) {
  safe_vector<FrameShiftResult> result;
  result.reserve(frameShiftResults.size());
  for (const auto& fsr : frameShiftResults) {
    result.push_back(fsr.frameShift);
  }
  return result;
}

PeptidesInternal translateGenes(                               //
  const NucleotideSequence& query,                             //
  const NucleotideSequence& ref,                               //
  const std::map<std::string, RefPeptideInternal>& refPeptides,//
  const GeneMap& geneMap,                                      //
  const safe_vector<int>& gapOpenCloseAA,                      //
  const NextalignOptions& options                              //
) {

  NucleotideSequence newQueryMemory(ref.size(), Nucleotide::GAP);
  NucleotideSequenceSpan newQuery{newQueryMemory};

  const CoordinateMapper coordMap{ref};

  safe_vector<PeptideInternal> queryPeptides;
  queryPeptides.reserve(geneMap.size());

  Warnings warnings;

  // For each gene in the requested subset
  for (const auto& [geneName, gene] : geneMap) {
    const auto& refPeptide = mapFind(refPeptides, geneName);
    invariant(refPeptide.has_value());
    if (!refPeptide) {
      continue;
    }

    // TODO: can be done once during initialization
    auto extractRefGeneStatus = extractGeneQuery(ref, gene, coordMap);
    if (extractRefGeneStatus.status != Status::Success) {
      const auto message = *extractRefGeneStatus.error;
      warnings.inGenes.push_back(GeneWarning{.geneName = geneName, .message = message});
      continue;
    }

    auto extractQueryGeneStatus = extractGeneQuery(query, gene, coordMap);
    if (extractQueryGeneStatus.status != Status::Success) {
      const auto message = *extractQueryGeneStatus.error;
      warnings.inGenes.push_back(GeneWarning{.geneName = geneName, .message = message});
      continue;
    }

    if (gene.strand == "-") {
      reverseComplementInPlace(*extractRefGeneStatus.result);
      reverseComplementInPlace(*extractQueryGeneStatus.result);
    }

    auto& refGeneSeq = *extractRefGeneStatus.result;
    const auto refGapCounts = countGaps(refGeneSeq);

    auto& queryGeneSeq = *extractQueryGeneStatus.result;
    const auto queryGeneSize = safe_cast<int>(queryGeneSeq.size());

    const auto queryGapCounts = countGaps(queryGeneSeq);
    const bool queryIsAllGaps = queryGapCounts.total >= queryGeneSize;

    // Handle the case when the query gene is completely missing
    if (queryGeneSeq.empty() || queryIsAllGaps) {
      const auto message = fmt::format(
        "When processing gene \"{:s}\": The gene consists entirely from gaps. "
        "Note that this gene will not be included in the results of the sequence.",
        geneName);
      warnings.inGenes.push_back(GeneWarning{.geneName = geneName, .message = message});
      continue;
    }

    // Make sure subsequent gap stripping does not introduce frame shift
    protectFirstCodonInPlace(refGeneSeq);
    protectFirstCodonInPlace(queryGeneSeq);

    // NOTE: frame shift detection should be performed on unstripped genes
    const auto nucRelFrameShifts = detectFrameShifts(refGeneSeq, queryGeneSeq);
    const auto frameShiftResults = translateFrameShifts(queryGeneSeq, nucRelFrameShifts, coordMap, gene);

    maskNucFrameShiftsInPlace(queryGeneSeq, frameShiftResults);

    // Strip all GAP characters to "forget" gaps introduced during alignment
    removeGapsInPlace(queryGeneSeq);

    debug_trace("Translating gene '{:}'\n", geneName);
    const auto queryPeptide = translate(queryGeneSeq, options.translatePastStop);

    // Bail out if the peptide is empty
    if (queryPeptide.empty()) {
      const auto message = fmt::format(
        "When processing gene \"{:s}\". Translated peptide is empty. "
        "Note that this gene will not be included in the results of the sequence.",
        geneName);
      warnings.inGenes.push_back(GeneWarning{.geneName = geneName, .message = message});
      continue;
    }

    debug_trace("Aligning peptide '{:}'\n", geneName);
    const AlignmentParams alignmentParams = calculateAaAlignmentParams(queryGapCounts, refGapCounts);
    const auto geneAlignmentStatus = alignPairwise(queryPeptide, refPeptide->peptide, gapOpenCloseAA, options.alignment,
      alignmentParams.bandWidth, alignmentParams.shift);

    if (geneAlignmentStatus.status != Status::Success) {
      const auto message = fmt::format(
        "When processing gene \"{:s}\": {:>16s}. "
        "Note that this gene will not be included in the results "
        "of the sequence.",
        geneName, *geneAlignmentStatus.error);
      warnings.inGenes.push_back(GeneWarning{.geneName = geneName, .message = message});
      continue;
    }

    auto stripped = stripInsertions(geneAlignmentStatus.result->ref, geneAlignmentStatus.result->query);

    maskPeptideFrameShiftsInPlace(stripped.queryStripped, frameShiftResults);

    safe_vector<FrameShiftResult> frameShiftResultsFinal = toExternal(frameShiftResults);

    queryPeptides.emplace_back(PeptideInternal{
      .name = geneName,                                      //
      .seq = std::move(stripped.queryStripped),              //
      .insertions = std::move(stripped.insertions),          //
      .frameShiftResults = std::move(frameShiftResultsFinal),//
    });
  }

  return PeptidesInternal{
    .queryPeptides = queryPeptides,//
    .warnings = warnings,          //
  };
}

File: packages_rs/nextclade/src/translate/translate_genes.rs
use crate::align::align::align_aa;
use crate::align::insertions_strip::{insertions_strip, Insertion};
use crate::align::params::AlignPairwiseParams;
use crate::align::remove_gaps::remove_gaps_in_place;
use crate::analyze::count_gaps::GapCounts;
use crate::gene::gene::Gene;
use crate::io::aa::Aa;
use crate::io::gene_map::GeneMap;
use crate::io::letter::{serde_deserialize_seq, serde_serialize_seq, Letter};
use crate::io::nuc::Nuc;
use crate::translate::coord_map::CoordMap;
use crate::translate::frame_shifts_detect::frame_shifts_detect;
use crate::translate::frame_shifts_translate::{frame_shifts_transform_coordinates, FrameShift};
use crate::translate::translate::translate;
use crate::utils::collections::{first, last};
use crate::utils::range::Range;
use crate::{make_error, make_internal_report};
use eyre::Report;
use indexmap::IndexMap;
use itertools::Itertools;
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

/// Results of the translation
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Translation {
  pub gene_name: String,
  #[serde(serialize_with = "serde_serialize_seq")]
  #[serde(deserialize_with = "serde_deserialize_seq")]
  pub seq: Vec<Aa>,
  pub insertions: Vec<Insertion<Aa>>,
  pub frame_shifts: Vec<FrameShift>,
  pub alignment_range: Range,
}

pub type TranslationMap = BTreeMap<String, Translation>;

/// Results of the aminoacid alignment parameters estimation
pub struct PeptideAlignmentParams {
  band_width: usize,
  mean_shift: i32,
}

/// Deduces aminoacid alignment params from nucleotide alignment.
/// Gap counts from nucleotide alignment are used to estimate rough parameters. This is by contrast to nucleotide
/// alignment, where these parameters are estimated using seed matching.
pub const fn calculate_aa_alignment_params(qry_gaps: &GapCounts, ref_gaps: &GapCounts) -> PeptideAlignmentParams {
  const BASE_BAND_WIDTH: usize = 5; // An arbitrary magic number to give some additional room for alignment
  let band_width = (qry_gaps.internal + ref_gaps.internal) / 3 + BASE_BAND_WIDTH;
  let mean_shift =
    (qry_gaps.leading as i32 - ref_gaps.leading as i32) / 3 + (qry_gaps.internal as i32 - ref_gaps.internal as i32) / 6;
  PeptideAlignmentParams { band_width, mean_shift }
}

/// Replaces first and second gap in a not-all-gap triplet with `N`
pub fn protect_codon_in_place(triplet: &mut [Nuc]) {
  if triplet[0].is_gap() {
    triplet[0] = Nuc::N;

    if triplet[1].is_gap() {
      triplet[1] = Nuc::N;
    }
  }
}

/// Finds the first non-GAP nucleotide and replaces GAPs in the corresponding triplet with Ns,
/// so that it's not getting stripped during subsequent gap stripping. This is to ensure the first codon is complete
/// and no shift is introduced during subsequent gap-stripping
pub fn protect_first_codon_in_place(seq: &mut [Nuc]) {
  // TODO: this is a very C-like function. Try to replace the indexed loop with something more idiomatic
  let length = seq.len();
  let end = length - (length % 3);
  for i in 0..end {
    if !seq[i].is_gap() {
      let triplet_begin = i - (i % 3);
      let triplet = &mut seq[triplet_begin..triplet_begin + 3];
      protect_codon_in_place(triplet);
      break;
    }
  }
}

/// Masks frame shifts in the nucleotide sequence with `N`
pub fn mask_nuc_frame_shifts_in_place(seq: &mut [Nuc], frame_shifts: &[FrameShift]) {
  // TODO: this is a very C-like function. Try to replace the indexed loop with something more idiomatic
  for frame_shift in frame_shifts {
    let mut current = frame_shift.nuc_rel.begin;
    let end = frame_shift.nuc_rel.end;
    while current < end {
      if !seq[current].is_gap() {
        seq[current] = Nuc::N;
      }
      current += 1;
    }
  }
}

pub fn fill_range_inplace(seq: &mut [Aa], range: &Range, letter: Aa) {
  // TODO: this is a very C-like function. Try to replace the indexed loop with something more idiomatic
  let mut current = range.begin;
  let end = range.end;
  while current < end {
    seq[current] = letter;
    current += 1;
  }
}

/// Masks gaps in frame-shifted regions of the peptide.
/// The frame-shifted region is likely misaligned, so the gaps added during peptide alignment don't make sense
/// and we cover them with `X`.
pub fn mask_peptide_frame_shifts_in_place(seq: &mut [Aa], frame_shifts: &[FrameShift]) {
  for frame_shift in frame_shifts {
    fill_range_inplace(seq, &frame_shift.gaps_leading.codon, Aa::Gap);
    fill_range_inplace(seq, &frame_shift.codon, Aa::X);
    fill_range_inplace(seq, &frame_shift.gaps_trailing.codon, Aa::Gap);
  }
}

pub fn translate_gene(
  qry_seq: &[Nuc],
  ref_seq: &[Nuc],
  gene: &Gene,
  ref_peptide: &Translation,
  gap_open_close_aa: &[i32],
  coord_map: &CoordMap,
  params: &AlignPairwiseParams,
) -> Result<Translation, Report> {
  let mut ref_gene_seq = coord_map.extract_gene(ref_seq, gene);
  let mut qry_gene_seq = coord_map.extract_gene(qry_seq, gene);

  let ref_gaps = GapCounts::new(&ref_gene_seq);
  let qry_gaps = GapCounts::new(&qry_gene_seq);

  if ref_gene_seq.is_empty() || qry_gaps.is_all_gaps() {
    return make_error!(
      "When processing gene \"{}\": The extracted gene sequence is empty or consists entirely from gaps",
      &gene.gene_name
    );
  }

  // If start and end nucs of qry are gaps, don't penalize them in alignment
  // TODO: Think about qry insertions, they will also be free?
  let aa_params = AlignPairwiseParams {
    // Set to false for internal genes
    left_terminal_gaps_free: first(&qry_gene_seq)?.is_gap(),
    right_terminal_gaps_free: last(&qry_gene_seq)?.is_gap(),
    ..*params
  };

  // Make sure subsequent gap stripping does not introduce frame shift
  protect_first_codon_in_place(&mut ref_gene_seq);
  protect_first_codon_in_place(&mut qry_gene_seq);

  // NOTE: frame shift detection should be performed on unstripped genes
  let nuc_rel_frame_shifts = frame_shifts_detect(&qry_gene_seq, &ref_gene_seq);
  let frame_shifts = frame_shifts_transform_coordinates(&nuc_rel_frame_shifts, &qry_gene_seq, coord_map, gene);

  mask_nuc_frame_shifts_in_place(&mut qry_gene_seq, &frame_shifts);

  // Strip all GAP characters to "forget" gaps introduced during alignment
  remove_gaps_in_place(&mut qry_gene_seq);

  let query_peptide = translate(&qry_gene_seq, gene, params)?;

  // Instead of performing seed matching, like we do for nucleotide alignment, here we estimate parameters
  // by counting gaps in the aligned nucleotide sequences;
  let PeptideAlignmentParams { band_width, mean_shift } = calculate_aa_alignment_params(&qry_gaps, &ref_gaps);

  let alignment = align_aa(
    &query_peptide.seq,
    &ref_peptide.seq,
    gap_open_close_aa,
    &aa_params,
    band_width,
    mean_shift,
  );

  let mut stripped = insertions_strip(&alignment.qry_seq, &alignment.ref_seq);

  mask_peptide_frame_shifts_in_place(&mut stripped.qry_seq, &frame_shifts);

  Ok(Translation {
    gene_name: gene.gene_name.clone(),
    seq: stripped.qry_seq,
    insertions: stripped.insertions,
    frame_shifts,
    alignment_range: Range::default(),
  })
}

/// Translates all requested genes
///
/// NOTE: we handle translation errors as warnings, so we return a collection of `Results` as is, to handle elsewhere
pub fn translate_genes(
  qry_seq: &[Nuc],
  ref_seq: &[Nuc],
  ref_peptides: &TranslationMap,
  gene_map: &GeneMap,
  coord_map: &CoordMap,
  gap_open_close_aa: &[i32],
  params: &AlignPairwiseParams,
) -> Result<IndexMap<String, Result<Translation, Report>>, Report> {
  gene_map
    .iter()
    .map(
      |(gene_name, gene)| -> Result<(String, Result<Translation, Report>), Report> {
        let ref_peptide = ref_peptides.get(&gene.gene_name).ok_or(make_internal_report!(
          "Reference peptide not found for gene {}",
          &gene.gene_name
        ))?;

        let res = translate_gene(
          qry_seq,
          ref_seq,
          gene,
          ref_peptide,
          gap_open_close_aa,
          coord_map,
          params,
        );

        Ok((gene_name.clone(), res))
      },
    )
    .collect::<Result<IndexMap<String, Result<Translation, Report>>, Report>>()
}
